playground/src/auto-importer/auto-importer.js

```diff
@@ -0,0 +1,245 @@
+      }
+    }
+  }
+
+  _fetchComponents() {
+    if (!Object.keys(COMPONENT_VERSION_MAP).length) {
+      let request = new Request(COMPONENTS_URL, {
+        headers: new Headers({'x-requested-by': 'playground'})
+      });
+      fetch(request)
+          .then((r) => r.json())
+          .then((data) => {
+            COMPONENT_VERSION_MAP = data;
+          });
+    }
+  }
+
+  /**
+   * Inserts missing <script> tags into the document, based on results from the
+   * AMP validator, and results from inspecting the current <head> structure.
+   *
+   * @param {!Object} missing The results from {@code _getMissingElements()}.
+   * @param {!Object} existing The results from {@code _parseHeadTag()}.
+   */
+  _insertMissingElements(missing, existing) {
+    const pos = existing.baseScriptTagEnd || existing.lastTag;
+    if (pos) {
+      const toAdd = Object.keys(missing.missingTags)
+          // Verify that all components to insert don't already exist: In some
+          // circumstances the validator has reported tags missing when in fact
+          // they are present.
+          .filter((e) => !existing.tags[e])
+          .map((e) => this._createAmpComponentElement(e));
+      if (missing.missingBaseScriptTag && !existing.baseScriptTagEnd) {
+        toAdd.unshift(AMP_ENGINE_SCRIPT_TAG);
+      }
+      const indented = toAdd.map((e) => ' '.repeat(existing.indent || 0) + e);
+      const cur = this.editor.getCursor();
+      this.editor.replaceRange('\n' + indented.join('\n'), pos, pos);
+      this.editor.setCursor(cur.line + indented.length, cur.ch);
+    }
+  }
+
+  /**
+   * Retrieves details of missing AMP components from the AMP validator results.
+   *
+   * @return {{
+   *   missingTags: Object,<string, number>,
+   *   missingBaseScriptTag: boolean
+   * }}
+   *
+   * missingTags: The keys of this dictionary are those missing AMP components.
+   * missingBaseScriptTag: True if the AMP engine <script> tag is not present.
+   */
+  _getMissingElements(validationResult) {
+    let missingComponentsSet = {};
+    let missingMandatoryAmpEngine = false;
+
+    for (let err of validationResult.errors) {
+      if (err.category === 'MANDATORY_AMP_TAG_MISSING_OR_INCORRECT') {
+        switch(err.code) {
+          case 'MANDATORY_TAG_MISSING':
+            if (err.params && err.params[0] === MISSING_ENGINE_PARAM) {
+              missingMandatoryAmpEngine = true;
+            }
+            break;
+          case 'MISSING_REQUIRED_EXTENSION':
+          case 'ATTR_MISSING_REQUIRED_EXTENSION':
+            if (err.params && err.params.length > 1) {
+              const tagName = err.params[1];
+              if (COMPONENT_VERSION_MAP[tagName]) {
+                missingComponentsSet[tagName] = 1;
+              } else {
+                console.log(`Warning: Unknown AMP component : ${tagName}`);
+              }
+            }
+            break;
+          default:
+            // no default
+        }
+      }
+    }
+    return {
```

Comment:
suggestion: create the return obj at the beginning of the method

```
const missingElements = {
  components: {},
  ampRuntime: false
}
```

Reply:
Done